src/shared/format-table.c

```diff
@@ -2916,40 +2945,43 @@ static int table_to_json_vertical(Table *t, JsonVariant **ret) {
         if (t->n_columns != 2)
                 return -EINVAL;
 
+        if (!ret)
+                return -EINVAL;
+
         /* Ensure we have no incomplete rows */
         assert(t->n_cells % t->n_columns == 0);
 
-        elements = new0(JsonVariant *, t->n_cells);
-        if (!elements)
-                return -ENOMEM;
-
-        CLEANUP_ARRAY(elements, n_elements, json_variant_unref_many);
-
-        for (size_t i = t->n_columns; i < t->n_cells; i++) {
+        for (size_t i = t->n_columns; i < t->n_cells; i += 2) {
+                _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+                _cleanup_free_ char *mangled = NULL;
+                const char *n;
 
-                if (i % t->n_columns == 0) {
-                        _cleanup_free_ char *mangled = NULL;
-                        const char *n;
+                n = table_get_json_field_name(t, i / t->n_columns - 1);
+                if (!n) {
+                        r = table_make_json_field_name(t, ASSERT_PTR(t->data[i]), &mangled);
+                        if (r < 0)
+                                return r;
 
-                        n = table_get_json_field_name(t, i / t->n_columns - 1);
-                        if (!n) {
-                                r = table_make_json_field_name(t, ASSERT_PTR(t->data[i]), &mangled);
-                                if (r < 0)
-                                        return r;
+                        n = mangled;
+                }
 
-                                n = mangled;
-                        }
+                r = table_data_to_json(t->data[i + 1], &v);
+                if (r < 0)
+                        return r;
 
-                        r = json_variant_new_string(elements + n_elements, n);
-                } else
-                        r = table_data_to_json(t->data[i], elements + n_elements);
+                r = json_variant_set_field(&o, n, v);
                 if (r < 0)
                         return r;
 
-                n_elements++;
+                if (typesafe_bsearch(&(size_t) { i / 2 }, t->delimiter_map, t->n_delimiter_map, row_compare)) {
+                        r = json_variant_append_array(&a, TAKE_PTR(o));
+                        if (r < 0)
+                                return r;
+                }
         }
 
-        return json_variant_new_object(ret, elements, n_elements);
+        *ret = a ? TAKE_PTR(a) : TAKE_PTR(o);
+        return 0;
```

Comment:
hmm, i wonder if outputting an array is really the right approach here. if you have a huge number of elements you stream out, it might be nicer to generate json-seq instead, because it means you can output/parse this stuff piecemeal and not keep the whole array in memory.